src/ng/directive/ngRepeat.js

```diff
@@ -425,126 +419,129 @@ var ngRepeatDirective = ['$parse', '$animate', '$compile', function($parse, $ani
 
-              nextNode = previousNode;
+            if (lastBlockMap[nextKey]) {
+              // we have already seen this object and need to reuse the associated scope/element
+              block = lastBlockMap[nextKey];
 
-              // skip nodes that are already pending removal via leave animation
-              do {
-                nextNode = nextNode.nextSibling;
-              } while (nextNode && nextNode[NG_REMOVED]);
+              // move
+              if (lastBlockMap[nextKey].index !== nextBlockMap[nextKey].index) {
+                // If this block has moved because the last previous block was removed,
+                // then use the last previous block to set previousNode.
+                lastKey = lastBlockOrder[lastBlockMap[nextKey].index - 1];
+                if (lastKey && !nextBlockMap[lastKey]) {
+                  previousNode = getBlockEnd(lastBlockMap[lastKey]);
+                }
 
-              if (getBlockStart(block) !== nextNode) {
-                // existing item which got moved
                 $animate.move(getBlockNodes(block.clone), null, previousNode);
+                block.index = nextBlockMap[nextKey].index;
               }
+
+              updateScope(block.scope, index,
+                valueIdentifier, nextBlockMap[nextKey].value,
+                keyIdentifier, nextBlockMap[nextKey].key, nextLength);
+
+              nextBlockMap[nextKey] = block;
               previousNode = getBlockEnd(block);
-              updateScope(block.scope, index, valueIdentifier, value, keyIdentifier, key, collectionLength);
+
             } else {
+              // enter
               // new item which we don't know about
               $transclude(function ngRepeatTransclude(clone, scope) {
-                block.scope = scope;
+                nextBlockMap[nextKey].scope = scope;
                 // http://jsperf.com/clone-vs-createcomment
                 var endNode = ngRepeatEndComment.cloneNode(false);
                 clone[clone.length++] = endNode;
 
                 $animate.enter(clone, null, previousNode);
                 previousNode = endNode;
+
                 // Note: We only need the first/last node of the cloned nodes.
                 // However, we need to keep the reference to the jqlite wrapper as it might be changed later
                 // by a directive with templateUrl when its template arrives.
-                block.clone = clone;
-                nextBlockMap[block.id] = block;
-                updateScope(block.scope, index, valueIdentifier, value, keyIdentifier, key, collectionLength);
+                nextBlockMap[nextKey].clone = clone;
+                updateScope(scope, nextBlockMap[nextKey].index,
+                  valueIdentifier, nextBlockMap[nextKey].value,
+                  keyIdentifier, nextBlockMap[nextKey].key, nextLength);
+
+                delete nextBlockMap[nextKey].key;
+                delete nextBlockMap[nextKey].value;
```

Comment:
This is new, right? Is this necessary? The tests do not fail without this.
 If we are doing this for every item, this might actually impact performance.